WIP Banded alignment.

rrahn opened the pull request:
Adds the banded alignment.

 * [x] Clean commit history

h-2 commented on include/seqan3/alignment/pairwise/policy/banded_score_dp_matrix_policy.hpp:
```diff
@@ -0,0 +1,216 @@
+
+    //!\brief A constant value for simulating minus infinity
+    static constexpr std::tuple_element_t<0, cell_type> INF =
+        std::numeric_limits<std::tuple_element_t<0, cell_type>>::lowest() / 2;
```
why lowest / 2 ?

rrahn replied:
to protect against underflow.

h-2 replied:
you are decreasing the INF and therefore need more range? What if you use small ints and need the range?

rrahn replied:
That's the problem of the algorithm. But for the scalar mode, this will actually never happen, as 32 bit integer operations will be the most efficient operations on most of our target cpus, so it will always use this. And in the case of SIMD I can use saturated arithmetics so there will be differences.

rrahn replied:
And additionally, the value is never propagated through the matrix. It is only to make the banded code easier to implement. A better approximation would be lowest - gap_score. But this affects the interface of some functions and as explained above I don‘t see the use case as in scalar mode you won‘t compute with 8bit scores only and hope for a better performance.

h-2 replied:
:+1:

h-2 commented on include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp:
```diff
@@ -141,6 +141,53 @@ class find_optimum_policy
         }
     }
+
+    /*!\brief Balances the total score of the alignment depending on the band settings and the alignment configuration.
```
What does it mean to "balance a score"? Sounds like getting revenge in Mafia slang :joy:

h-2 commented on include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp:
```diff
@@ -141,6 +141,53 @@ class find_optimum_policy
+     * \tparam dimension_type  The type of the matrix dimensions.
+     * \tparam band_type       The type of the band.
+     * \tparam gap_scheme_type The type of the gap_scheme.
+     * \param[in,out] total                  The total score to be updated.
```
indent

h-2 commented on include/seqan3/alignment/pairwise/policy/unbanded_dp_matrix_policy.hpp:
```diff
@@ -66,22 +68,32 @@ class unbanded_dp_matrix_policy
+    {
+        ++current_column_index;
+    }
+
```
this function is constexpr but the previous isn't. Is any of this supposed to run at compile-time? In any case it should be consistent, I think!

rrahn replied:
and yes if constexpr allocator come into the standard, it could be possible to align at compile time 😏

h-2 commented on test/unit/alignment/exception_test.cpp:
```diff
@@ -0,0 +1,37 @@
+TEST(alignment_exception, invalid_alignment_configuration)
+{
+
+    auto throw_lambda = []() {
```
to next line

h-2 commented on test/unit/alignment/exception_test.cpp:
```diff
@@ -0,0 +1,37 @@
+    catch (...)
+    {
+        FAIL();
+    }
```
what are you testing here? :thinking:

rrahn replied:
I am testing that the message is correctly passed to the exception object. And in the case of something completely unexpected it should fail.

h-2 replied:
You mean you wrote a 25 line testing for the one-line constructor of your exception that only forwards to a std::exception? Isn't that a little too much?
Couldn't you rather test a case where the exception is thrown by your algorithm?

rrahn replied:
One test per class.

h-2 commented on test/unit/alignment/pairwise/alignment_configurator_test.cpp:
```diff
@@ -0,0 +1,180 @@
+auto setup()
+{
+    auto data = std::tuple{"ACGT"_dna4, "ACGT"_dna4};
+    return ranges::view::single(std::move(data)) | view::persist;
```
local and global `data`? forgot to remove one?

h-2 commented on test/unit/alignment/pairwise/alignment_configurator_test.cpp:
```diff
@@ -0,0 +1,180 @@
+        auto cfg_upper = cfg_base | align_cfg::band{static_band{lower_bound{5}, upper_bound{6}}};
+
+        auto r = setup();
+        {
```
newline between scope and previous line if unconnected

h-2 commented on test/unit/alignment/pairwise/fixture/semi_global_affine_banded.hpp:
```diff
@@ -0,0 +1,207 @@
+
+#include "alignment_fixture.hpp"
+
+namespace seqan3::fixture::semi_global::affine::banded
```
:open_mouth:

rrahn replied:
it makes debugging the tests quite simple.

rrahn replied:
Or not debugging, but you see from the gtest error output which functions failed. And there will be a lot more of tests added in the future.

h-2 replied:
You really need a 5-fold nested namespace within the *test framework*?

rrahn replied:
Yes.

h-2 replied:
Can you at least put a `::test` before it to make clear that this is not part of the library?

rrahn replied:
Well, if it makes the context more clear, why not.

rrahn commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -153,14 +274,17 @@ class alignment_algorithm :
 
-        this->check_score_last_row(get<0>(*(seqan3::end(col) - 1)), get<3>(cache));
+        if constexpr (is_banded)
+            this->check_score_last_row(get<0>(get<0>(*(--seqan3::end(col)))), get<3>(cache));
```
Why would they not? This is specific to the algorithm implementation and not to any input from the user. prev it is then, but why would they not add front/back as well as customisation points? It seems like a very legit task on ranges to do.

rrahn replied:
Actually, we should use `ranges::prev` and `ranges::next` as these are then customisation point objects. I will add forwards to `std/iterator`

h-2 replied:
> Why would they not?

It's not covered  by the requirements. 

> Actually, we should use `ranges::prev` and `ranges::next`

Yes.

> ... as these are then customisation point objects. I will add forwards to `std/iterator`

Why, what do you want to customise?

h-2 replied:
> I will add forwards to `std/iterator`

I got this now, Yes, we need aliases.

rrahn replied:
> It's not covered by the requirements.

Whose requirements? If it makes you happy I can add a static assert?

h-2 replied:
It's not required by the concepts. We shouldn't write code that makes silent assumptions that some STL iterators do something that is not promised. Even if it works in the concrete case and on this platform, why should we not use the standard-conform method?

rrahn replied:
I still don' get it. Why ist the code not standard-conform? I agreed already on using `ranges::prev` here. But the assumption whether the range is bidirectional or not is completely independent from any user input. It is the invariant of the class that assures bidirectional ranges over the score/trace matrix. Note, that I never decrement something that is passed from outside. Or what did I miss? What do you think should be constrained exactly in this case?

h-2 replied:
With ranges::prev it is standards conform, sorry for the confusion!

rrahn commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -153,14 +274,17 @@ class alignment_algorithm :
+        ranges::for_each(col | ranges::view::drop_exactly(1), [&cache, this](auto && cell)
         {
-            this->init_column_cell(cell, cache);
+            this->init_column_cell(std::forward<decltype(cell)>(cell), cache);
```
Yet, always, as it is a zip view internally returning tuple of tuples.

h-2 replied:
A tuple of tuples isn't necessarily heap-allocated. In the case where it contains only builtins, it is not. And since this is "hot code" I am pretty sure copies would be faster, just saying...

rrahn replied:
I don't get this. Dereferencing the zip iterator will return a rvalue tuple. Now why would a copy be better than a perferctly forwarded rvalue? And if the compiler sees that eveything is a builtin he copies anyway.

h-2 replied:
> Now why would a copy be better than a perferctly forwarded rvalue?
> And if the compiler sees that eveything is a builtin he copies anyway.

Yes and no. I think it's still a difference whether you take by copy or take by &&. It will affect the compilers ability to predict aliasing or not?

Anyway, not so important...

rrahn replied:
I think the relevant performance difference (if any) can be evaluated with the simd code

rrahn commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -142,6 +152,117 @@ class alignment_algorithm :
+        static_assert(std::is_signed_v<diff_type>,  "Only signed types can be used to test the band parameters.");
+
+        if (static_cast<diff_type>(band.lower_bound) >
+            std::distance(seqan3::begin(first_batch), seqan3::end(first_batch)))
```
Negative bands are allowed as long as they intersect with the matrix. 
batch is a either a single range or a range of simd vector

h-2 replied:
> Negative bands are allowed as long as they intersect with the matrix.

But a negative number will never be bigger than a distance which is >= zero  :thinking: 

> batch is a always a single range over either a single alphabet or a simd vector

And this line covers both? Wouldn't you need to add a factor for the case of collection-of-simd-vectors?

rrahn replied:
upper and lower bound can be both negative, positive or negative and positive.
This line says if the value for the lower bound is bigger than the length of the sequence in the horizontal direction, the band is invalid. The simd sequence is as long as the longest sequence.
So for some sequences it could be invalid. But we deal with this when we need to. The other case (upper diagonal exceeds the length of the vertical sequence) is handled below.

h-2 replied:
> upper and lower bound can be both negative, positive or negative and positive.
> This line says if the value for the lower bound is bigger than the length of the sequence in the horizontal direction, the band is invalid.

Got this!

> The simd sequence is as long as the longest sequence.

I still don't understand how this function can handle batches of sequences, the distance will just return the size of the batch and not the size of the sequences :question:

rrahn replied:
Well, it depends on the behaviour we want. All smaller sequences will be padded with some extra symbols so we can compute the alignment and later it will be resolved to practically aligning seq1 to the empty seq and vice versa for seq2. Or we refactor it later to check the individual sequences and throw. But this comes anyway later.

h-2 replied:
can we talk about this on gitter or on the phone? I have the feeling I am missing something o_O

h-2 replied:
Ok, I got it now. 

As discussed in the chat I would hugely appreciate a rename to something normal like "rng_t" oder "sequence_t" with a description in the documentation that states something like `In the case of generic inter-sequence vectorisation this is a range over SIMD vectors instead of a range over alphabet letters.`
The regular ranges would be over the original seqan3::Alphabet but in SIMD case they would be over simd vectors of the ranks, right?

rrahn replied:
It's fine for me and yes the last statement is right

rrahn commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -118,14 +124,18 @@ class alignment_algorithm :
     template <std::ranges::ForwardRange first_batch_t, std::ranges::ForwardRange second_batch_t>
     auto operator()(first_batch_t const & first_batch, second_batch_t const & second_batch)
+    //!\cond
+        requires !is_banded
```
It is only a distinction for the banded version, as the kernel differs widely from the unbanded version and adding a lot of `constexpr ifs` would not make the code more readable.

rrahn commented on include/seqan3/alignment/pairwise/alignment_configurator.hpp:
```diff
@@ -257,14 +335,14 @@ constexpr function_wrapper_t alignment_configurator::configure_free_ends_initial
     {
         // Define the trait for the initialisation policy
-        struct _trait
+        struct _trait_type
```
They have fun?!

h-2 replied:
![](https://media.makeameme.org/created/are-you-5acbb2.jpg)

rrahn replied:
🤣

rrahn replied:
you really want me to change this?

h-2 replied:
well we don't use `_` anywhere except to mark "private" variables in aggregate types or in the case of name clashes of local variables. I don't see the necessity here to use it and it makes the names stand out needlessly..

rrahn replied:
For the lambda it is true. They are not seen outside of the scope, but with the traits type I wanted to avoid possible name clashes with function or class templates

rrahn replied:
And last time I complained about your `_char_type` thing you were like: "Na, this is just more readable." 
...
just saying 😉

h-2 replied:
Yeah, well I had two things there were 100% identical in all situations were code is generated so I thought it was important to give them very similar names. In this case I see no obvious reason for the underscore, that's why I am asking. Stuff is private anyway, it doesn't need to be private-private does it?

rrahn replied:
I changed it already

rrahn commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -142,6 +152,117 @@ class alignment_algorithm :
+    auto operator()(first_batch_t const & first_batch, second_batch_t const & second_batch)
+    //!\cond
+        requires is_banded
+    //!\endcond
```
It works well. But it might not clear where the destinction comes from. So probably the requires can be left inside.

h-2 replied:
yes, might be good to try that, usually it works for a simple term and will be included in the signature!

rrahn replied:
I fixed already. The view is outdated.

h-2 commented on include/seqan3/alignment/pairwise/alignment_algorithm.hpp:
```diff
@@ -142,6 +151,115 @@ class alignment_algorithm :
+        }
+
+        if (static_cast<diff_type>(std::abs(band.upper_bound)) >
+            std::distance(seqan3::begin(second_batch), seqan3::end(second_batch)))
```
Why do you use `std::abs` here now, shouldn't a negative value be kept as negative?

rrahn replied:
Because it now tests, whether the upper bound starts behind the end of the vertical sequence. This means the value is negative and needs to be compared with the distance type between begin and end.

```
 01234567
0
1
2
3
4
5
.
.x
. x
.  x
.x  x
. x   x
```

Here the band is set to lower_bound:-10 upper_bound: -7; Hence I use the abs to get the positive value.

h-2 replied:
Ok, I get the reason you need to change the sign, but should it then not be just `-band.upper_bound` and not abs?

rrahn replied:
They mean and do literally the same (I think even from the issued instructions) but the abs version is more expressive in what it does, I think.

h-2 commented:
BTW: our documentation still has nothing on the "align_pairwise" function which is the main user-visible entry point. Can you add this with this PR or put it somewhere up on you (long) TODO list? I think this is essential before the PMSB.

rrahn commented:
@h-2 the documentation of the alignment module is on the list for this sprint. It won't be part of this PR though

rrahn commented:
@h-2 I think i adressed now all of your issues. Can you please also have a look at `seqan3/std/iterator`, where I added the alias declarations for the iterator operations. I added this recently and it might have been remained undetected.
I did not add any documentation entry for it as they are also not yet part of cppreference.

rrahn commented:
@h-2 I think i addressed now all of your issues. Can you please also have a look at `seqan3/std/iterator`, where I added the alias declarations for the iterator operations. I added this recently and it might have been remained undetected.
I did not add any documentation since the documentation is also missing on  cppreference.

rrahn commented:
Ok, then I clean the commits and merge the pr. 🎆
